### DISCUSSION TITLE
Internalization support

### DISCUSSION THREAD

**Discussion by arvigeus** [epicweb-dev/epic-stack#44] [Ideas]
It's a common requirement for medium to large projects.

Pros:

- Saving users some decisions

Cons:

- Could be too complicated to implement, or not be that used.

**Comment by kentcdodds** [comment]
This is definitely planned

**Reply by kasperpeulen** [reply]
Would you accept a PR for it?

(Not sure if you rather do it yourself)

**Reply by kentcdodds** [reply]
An example is always the best first step!

**Comment by sergiodxa** [comment]
One nice thing about i18n is that if you do it from the beginning, even if you only use one locale, it makes it trivial to add more locales later (your texts are already prepared for that!) and if you use keys to identify texts then your messages files makes it easier to update the copy later without even changing the actual code.

**Comment by rodbs** [comment]
@kentcdodds  How are you planning to implement i18n in epic-stack? using `remix-i18next`? 
Isn't there a simpler way (for basic i18n support)? the mechanism is not similar to the one in theme switching (having a cookie to track the theme/locale server-side)?

**Reply by sergiodxa** [reply]
There's more in i18n than just storing the user locale somewhere (and you could even store it in the DB), you have to load translation files, pluralize messages, render the messages, sometimes render a message with JSX in the middle, all of that is handled by `i18next` and `react-i18next`.

The `remix-i18next` package gives you tools to know the locale from the request headers, Remix sessions and cookies, or the URL, and to use `i18next` server-side to translate texts in the loader.

A simpler approach could be as simple as importing a JSON with your translation messages in every language you support, store the user locale somewhere (DB, session, cookie, URL, etc.) and use the locale to grab the translated messages and render them, without any JSX integration and with manual pluralization.

**Reply by sergiodxa** [reply]
BTW I still think EpicStack should come with i18n support out of the box, even if most apps uses only English adding it later will be harder.

**Reply by kentcdodds** [reply]
> EpicStack should come with i18n support out of the box

I 100% agree

**Reply by koikar** [reply]
There is already a fantastic example of epic-stack with remix-i18next created by Remi @rperon 

Setting up the automatic parser is also essential for any real-world application. I am used to work with natural-language strings in the code base (coming from lingui.js) so I changed my config to support that.

Following Remi's guide together with @sergiodxa documentation you should be up an running in no time 👍

This was one of the first things I checked for in the epic-stack. So having it built-in would be great for all of us living/working outside the US.

https://github.com/rperon/epic-stack-with-i18n
